rest/route/host.go

```diff
@@ -314,6 +316,133 @@ func getLimit(vals url.Values) (int, error) {
 	return limit, nil
 }
 
+////////////////////////////////////////////////////////////////////////
+//
+// POST /users/{user_id}/clear_hosts
+
+func makeClearHostsByUser(sc data.Connector, env evergreen.Environment) gimlet.RouteHandler {
+	return &clearHostsHandler{
+		sc:  sc,
+		env: env,
+	}
+}
+
+type clearHostsHandler struct {
+	user   string
+	dryRun bool
+
+	env evergreen.Environment
+	sc  data.Connector
+}
+
+func (ch clearHostsHandler) Factory() gimlet.RouteHandler {
+	return &clearHostsHandler{
+		sc:  ch.sc,
+		env: ch.env,
+	}
+}
+
+func (ch *clearHostsHandler) Parse(ctx context.Context, r *http.Request) error {
+	ch.user = gimlet.GetVars(r)["user_id"]
+	if ch.user == "" {
+		return gimlet.ErrorResponse{
+			Message:    "user cannot be empty",
+			StatusCode: http.StatusBadRequest,
+		}
+	}
+	vals := r.URL.Query()
+	ch.dryRun = vals.Get("dry_run") == "true"
+
+	return nil
+}
+
+func (ch *clearHostsHandler) Run(ctx context.Context) gimlet.Responder {
+	usr := gimlet.GetUser(ctx)
+	opts := model.APIHostParams{
+		UserSpawned: true,
+	}
+	// returns all up-hosts for user
+	hosts, err := ch.sc.FindHostsInRange(opts, ch.user)
+	if err != nil {
+		return gimlet.NewJSONErrorResponse(errors.Wrap(err, "database error getting hosts"))
+	}
+
+	volumes, err := ch.sc.FindVolumesByUser(ch.user)
+	if err != nil {
+		return gimlet.NewJSONErrorResponse(errors.Wrap(err, "database error getting volumes"))
+	}
+
+	toTerminate := model.APIClearHostsResults{
+		Hosts:   []string{},
+		Volumes: []string{},
+	}
+	currentTime := time.Now()
+
+	mgrCache := map[cloud.ManagerOpts]cloud.Manager{}
```

Comment:
why is it important to keep a cache? Is it expensive to get a manager?

Reply:
Not necessarily but it could be. I just don't see the point in doing duplicate work: for example if all your hosts are us-east-1 and ec2OnDemand then you only need the one manager.

Reply:
but is the cache lookup less work than `GetManager`?

Reply:
Well I would say so, theoretically hash map lookup is O(1) and GetManager potentially makes API calls.